File: go/example_code/cloudtrail/create_trail.go
// snippet-comment:[These are tags for the AWS doc team's sample catalog. Do not remove.]
// snippet-sourceauthor:[Doug-AWS]
// snippet-sourcedescription:[create_trail.go demonstrates how to create a CloudTrail trail.]
// snippet-keyword:[Amazon CloudTrail]
// snippet-keyword:[CreateTrail function]
// snippet-keyword:[Go]
// snippet-sourcesyntax:[go]
// snippet-service:[cloudtrail]
// snippet-keyword:[Code Sample]
// snippet-sourcetype:[full-example]
// snippet-sourcedate:[2020-1-6]
/*
   Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
// snippet-start:[cloudtrail.go.create_trail.complete]
package main

// snippet-start:[cloudtrail.go.create_trail.imports]
import (
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/cloudtrail"
    "github.com/aws/aws-sdk-go/service/s3"
    "github.com/aws/aws-sdk-go/service/sts"

    "encoding/json"
    "flag"
    "fmt"
)

// snippet-end:[cloudtrail.go.create_trail.imports]

func main() {
    // Get required trail name and bucket name
    // snippet-start:[cloudtrail.go.create_trail.vars]
    trailNamePtr := flag.String("n", "", "The name of the trail")
    bucketNamePtr := flag.String("b", "", "the name of the bucket to which the trails are uploaded")
    addPolicyPtr := flag.Bool("p", false, "Whether to add the CloudTrail policy to the bucket")

    flag.Parse()

    if *trailNamePtr == "" || *bucketNamePtr == "" {
        fmt.Println("You must supply a trail name and bucket name.")
        return
    }
    // snippet-end:[cloudtrail.go.create_trail.vars]

    // Initialize a session that the SDK will use to load
    // credentials from the shared credentials file ~/.aws/credentials.
    // snippet-start:[cloudtrail.go.create_trail.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))
    // snippet-end:[cloudtrail.go.create_trail.session]

    // snippet-start:[cloudtrail.go.create_trail.policy]
    if *addPolicyPtr {
        svc := sts.New(sess)
        input := &sts.GetCallerIdentityInput{}

        result, err := svc.GetCallerIdentity(input)
        if err != nil {
            fmt.Println("Got error snarfing caller identity:")
            fmt.Println(err.Error())
            return
        }

        accountID := aws.StringValue(result.Account)

        s3Policy := map[string]interface{}{
            "Version": "2012-10-17",
            "Statement": []map[string]interface{}{
                {
                    "Sid":    "AWSCloudTrailAclCheck20150319",
                    "Effect": "Allow",
                    "Principal": map[string]interface{}{
                        "Service": "cloudtrail.amazonaws.com",
                    },
                    "Action":   "s3:GetBucketAcl",
                    "Resource": "arn:aws:s3:::" + *bucketNamePtr,
                },
                {
                    "Sid":    "AWSCloudTrailWrite20150319",
                    "Effect": "Allow",
                    "Principal": map[string]interface{}{
                        "Service": "cloudtrail.amazonaws.com",
                    },
                    "Action":   "s3:PutObject",
                    "Resource": "arn:aws:s3:::" + *bucketNamePtr + "/AWSLogs/" + accountID + "/*",
                    "Condition": map[string]interface{}{
                        "StringEquals": map[string]interface{}{
                            "s3:x-amz-acl": "bucket-owner-full-control",
                        },
                    },
                },
            },
        }

        policy, err := json.Marshal(s3Policy)
        if err != nil {
            fmt.Println("Error marshalling request")
            return
        }

        // Create S3 service
        s3Svc := s3.New(sess)

        // Now set the policy on the bucket
        _, err = s3Svc.PutBucketPolicy(&s3.PutBucketPolicyInput{
            Bucket: aws.String(*bucketNamePtr),
            Policy: aws.String(string(policy)),
        })
        if err != nil {
            fmt.Print("Got error adding bucket policy:")
            fmt.Print(err.Error())
            return
        }

        fmt.Printf("Successfully set bucket %q's policy\n", *bucketNamePtr)
    }
    // snippet-end:[cloudtrail.go.create_trail.policy]

    // snippet-start:[cloudtrail.go.create_trail.create]
    svc := cloudtrail.New(sess)

    input := &cloudtrail.CreateTrailInput{
        Name:         aws.String(*trailNamePtr),
        S3BucketName: aws.String(*bucketNamePtr),
    }

    _, err := svc.CreateTrail(input)
    if err != nil {
        fmt.Println("Got error calling CreateTrail:")
        fmt.Println(err.Error())
        return
    }

    fmt.Println("Created the trail", *trailNamePtr, "for bucket", *bucketNamePtr)
    // snippet-end:[cloudtrail.go.create_trail.create]
}

// snippet-end:[cloudtrail.go.create_trail.complete]

File: ruby/example_code/cloudtrail/create_trail.rb
# frozen_string_literal: true

# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

# Demonstrates how to create an AWS CloudTrail trail

# snippet-start:[cloudtrail.Ruby.createTrail]
require "aws-sdk-cloudtrail" # v2: require 'aws-sdk'
require "aws-sdk-s3"
require "aws-sdk-sts"

def create_trail_example(s3_client, sts_client, cloudtrail_client, trail_name, bucket_name)

  resp = sts_client.get_caller_identity({})
  account_id = resp.account

  # Attach policy to an Amazon Simple Storage Service (S3) bucket.
  s3_client.create_bucket(bucket: bucket_name)
  begin
    policy = {
      "Version" => "2012-10-17",
      "Statement" => [
        {
          "Sid" => "AWSCloudTrailAclCheck20150319",
          "Effect" => "Allow",
          "Principal" => {
            "Service" => "cloudtrail.amazonaws.com"
          },
          "Action" => "s3:GetBucketAcl",
          "Resource" => "arn:aws:s3:::#{bucket_name}"
        },
        {
          "Sid" => "AWSCloudTrailWrite20150319",
          "Effect" => "Allow",
          "Principal" => {
            "Service" => "cloudtrail.amazonaws.com"
          },
          "Action" => "s3:PutObject",
          "Resource" => "arn:aws:s3:::#{bucket_name}/AWSLogs/#{account_id}/*",
          "Condition" => {
            "StringEquals" => {
              "s3:x-amz-acl" => "bucket-owner-full-control"
            }
          }
        }
      ]
    }.to_json

    s3_client.put_bucket_policy(
      bucket: bucket_name,
      policy:
    )
    puts "Successfully added policy to bucket #{bucket_name}"
  end

  begin
    cloudtrail_client.create_trail({
                                      name: trail_name, # required
                                      s3_bucket_name: bucket_name # required
                                    })

    puts "Successfully created trail: #{trail_name}."
  rescue StandardError => e
    puts "Got error trying to create trail #{trail_name}:\n #{e}"
    puts e
    exit 1
  end
  # snippet-end:[cloudtrail.Ruby.createTrail]
  "test"
end

if __FILE__ == $0
  @s3_client = Aws::S3::Client.new
  @sts_client = Aws::STS::Client.new
  @cloudtrail_client = Aws::CloudTrail::Client.new
  trail_name = "example-code-trail-#{rand(10**4)}"
  bucket_name = "example-code-bucket-#{rand(10**4)}"
  create_trail_example(@s3_client, @sts_client, @cloudtrail_client, trail_name, bucket_name)
end
